File: 85-maximal-rectangle/85-maximal-rectangle.py
class Solution(object):
    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        nrow, ncol = len(matrix), len(matrix[0])
        def maxArea(heights):
            stack = [-1]
            res = 0
            for i in range(ncol):
               # top = heights[stack[-1]]
                while(stack[-1] != -1 and heights[stack[-1]] > heights[i] ):
                    popped = stack.pop()
                    width = i - stack[-1] - 1
                    height = heights[popped]
                    res = max(res,width*height)
                stack.append(i)
            while(stack[-1] != -1):
                popped = stack.pop()
                width = ncol - stack[-1] - 1
                height = heights[popped]
                res = max(res,width*height)   
            return res
        final = 0
        heights = []
        for i in range(nrow):
            for j in range(ncol):
                if len(heights) < ncol:
                    if matrix[i][j] == "1":
                        heights.append(1)
                    else:
                        heights.append(0)
                elif matrix[i][j] == "0":
                    heights[j] = 0
                else:
                    heights[j] += 1
           # print(heights)
            final = max(maxArea(heights),final)
        return final
            
                
                    
                        
        
                            
                        
                    
        

File: 85-maximal-rectangle/85-maximal-rectangle.java
class Solution {
    int nrow;
    int ncol;
    public int maximalRectangle(char[][] matrix) {
        nrow = matrix.length; ncol = matrix[0].length;
        int[][] numOfOne = new int[nrow][ncol];
        countOne(numOfOne, matrix);
        int ans = 0;
        for(int[] array : numOfOne){
            ans = Math.max(ans, findLargestRec(array));
        }
        return ans;
        
        
        
    }
    public void countOne(int[][] numOfOne, char[][] matrix){
        for(int row = 0; row < nrow; row++)
            for(int col = 0; col < ncol; col++){
                char curChar = matrix[row][col];
                if(curChar == '1'){
                    numOfOne[row][col] += 1;
                }
                if(row > 0 && matrix[row-1][col] == '1' && matrix[row][col] == '1'){
                    numOfOne[row][col] += numOfOne[row - 1][col];
                }
            }
    }
    
    public int findLargestRec(int[] heights){
     //   System.out.printf("%d %d\n", heights[0], heights[1]);
        //leftLargest
        LinkedList<int[]> leftStack = new LinkedList<>();
        int N = heights.length;
        int[] leftSmallest = new int[N];
        int[] rightSmallest = new int[N];
        
        for(int i = 0; i < N; i++){
            int height = heights[i];
            rightSmallest[i] = N;
            //left largest
            while(!leftStack.isEmpty() && leftStack.peekLast()[0] >= height){
                int[] popped = leftStack.pollLast();
                rightSmallest[popped[1]] = i;
                }
                if(leftStack.isEmpty()){
                    leftSmallest[i] = -1;
                }else{
                    leftSmallest[i] = leftStack.peekLast()[1];
                }
            leftStack.addLast(new int[]{height, i});
        }
        int ans = 0;
        for(int j = 0; j < N; j++){
            int curHeight = heights[j];
            int left = leftSmallest[j], right = rightSmallest[j];
          //  System.out.printf("%d %d\n", left, right);
            if(left != j){
                left = left + 1;
            }
            if(right != j){
                right = right - 1;
            }
            ans = Math.max(ans, (right - left + 1) * curHeight);
        }
        return ans;      
    }
        
        
        
        
    
}
